### ISSUE TITLE
Amethyst 0.15.6 does not work at all (0.15.4 is fine)

### ISSUE THREAD

**Issue by ajorpheus** [ianyh/Amethyst#1240]
**Describe the bug**
Amethyst v0.15.6 does not work at all ( None of the keyboard shortcuts work, none of the layouts take effect etc).

**Applications:**
All

**To Reproduce**
1. Cleanly uninstall v0.15.4 removing the state files and preferences (`~/Library/Preferences/com.amethyst.Amethyst.plist` ,  `~/Library/Caches/com.amethyst.Amethyst`)
2. Remove Amethyst from Security and Privacy --> Accessibility
3. Install v0.15.6 of Amethyst (manually using zip downloaded from Github releases page)
4. Start Amethyst v0.15.6 for the first time
5. Prompted to enable Accessibility permissions, do so. Relaunch Amethyst
6. Record a shortcut to Restart Amethyst. Use shortcut. 

**Expected behavior**
Use the shortcut above, Amethyst is not relaunched.  This is just one of the problems, basically Amethyst doesn't do anything.

**Screenshots**
N/A

**Versions:**
 - macOS: 10.15.7
 - Amethyst: v0.15.6 (93)
 
**Debug Info**
```
$ /Applications/Amethyst.app/Contents/MacOS/Amethyst --debug-info
Version: 0.15.6 (93)

OS version: Version 10.15.7 (Build 19H2)

Screens:
	(0.0, 0.0, 2560.0, 1440.0) [(0.0, 0.0, 2560.0, 1440.0)]

Configuration:
window-margins: 0
window-resize-step: 5
mod1: (
    option,
    shift
)
screen-padding-right: 0
use-canary-build: 0
enables-layout-hud-on-space-change: 1
restore-layouts-on-launch: 1
new-windows-to-main: 0
screen-padding-left: 0
window-minimum-width: 0
mod2: (
    option,
    shift,
    control
)
screen-padding-top: 0
debug-layout-info: 0
focus-follows-mouse: 0
layouts: (
    tall,
    wide,
    fullscreen,
    column
)
screen-padding-bottom: 0
ignore-menu-bar: 0
float-small-windows: 1
window-margin-size: 0
mouse-follows-focus: 0
enables-layout-hud: 1
floating: (
)
window-minimum-height: 0
floating-is-blacklist: 1
follow-space-thrown-windows: 1

```

**Additional context**
This has been a problem since 0.15.6 earliest beta as well.

**Comment by mcranieri** [comment]
It's happening also for OSX 12.3.1 (Build 21E258)

```
Version: 0.15.6 (93)

OS version: Version 12.3.1 (Build 21E258)

Screens:
        (0.0, 0.0, 1512.0, 982.0) [(0.0, 0.0, 1512.0, 982.0)]
        (1512.0, -458.0, 3440.0, 1440.0) [(1512.0, 0.0, 3440.0, 1440.0)]

Configuration:
screen-padding-left: 0
enables-layout-hud: 1
ignore-menu-bar: 0
follow-space-thrown-windows: 1
mod1: (
    option,
    shift
)
screen-padding-bottom: 0
window-margins: 0
screen-padding-top: 0
mod2: (
    option,
    shift,
    control
)
floating: (
)
window-resize-step: 5
window-margin-size: 0
floating-is-blacklist: 1
use-canary-build: 0
focus-follows-mouse: 0
restore-layouts-on-launch: 1
window-minimum-width: 0
window-minimum-height: 0
new-windows-to-main: 1
float-small-windows: 1
debug-layout-info: 0
enables-layout-hud-on-space-change: 1
layouts: (
    "middle-wide",
    wide,
    fullscreen
)
mouse-follows-focus: 0
screen-padding-right: 0
```

**Comment by ajorpheus** [comment]
@ianyh  Just wanted to add that v0.15.5 works far better than 0.15.6 (though not as well as 0.15.4, it seems to have trouble getting binary windows partitioning right and I always have to relaunch Amethyst after launching new windows for 0.15.5 to get it right).

So it is definitely one of commits in https://github.com/ianyh/Amethyst/compare/v0.15.5...v0.15.6  that is the culprit.

**Comment by ajorpheus** [comment]
Ok, so I played around with this a bit more. I suspect this has something to do with the 'Floating' preferences (and possibly the commit that added the regex support to Window title matching).

I added a `.*` and changed the selection to 'Automatically float applications listed' and then I deleted the regex. After a relaunch, Amethyst 0.15.6 began working again. Very weird.

**Comment by ajorpheus** [comment]
Ah, and it has stopped working again today. 
v0.15.6 has completely stopped working now, and I can see some errors in `Console` 👍 

```
default	11:04:02.799417+0100	Amethyst	FRONTLOGGING: version 1
default	11:04:02.799443+0100	Amethyst	Registering, pid=43913
default	11:04:02.800772+0100	Amethyst	CHECKIN: pid=43913
default	11:04:02.814129+0100	Amethyst	CHECKEDIN: pid=43913 asn=0x0-0x2c82c8 foreground=0
default	11:04:02.887854+0100	Amethyst	Registered, pid=43913 ASN=0x0,0x2c82c8
default	11:04:02.887974+0100	Amethyst	Registered, pid=43913 cgConnectionID=c5827
default	11:04:02.889174+0100	Amethyst	BringForward: pid=43913 asn=0x0-0x2c82c8 bringForward=0 foreground=0 uiElement=1 launchedByLS=1 modifiersCount=1 allDisabled=0
default	11:04:02.889214+0100	Amethyst	BringFrontModifier: pid=43913 asn=0x0-0x2c82c8 Modifier 0 hideAfter=0 hideOthers=0 dontMakeFrontmost=0 mouseDown=0/0 seed=0/0
default	11:04:02.891904+0100	Amethyst	Current system appearance, (HLTB: 1), (SLS: 0)
default	11:04:02.897026+0100	Amethyst	Post-registration system appearance: (HLTB: 1)
default	11:04:02.922457+0100	Amethyst	NSApp cache appearance:
-NSRequiresAquaSystemAppearance: 0
-appearance: (null)
-effectiveAppearance: <NSCompositeAppearance: 0x6000008a1680
 (
    "<NSAquaAppearance: 0x6000008ad700>",
    "<NSAquaAppearance: 0x6000008ad600>",
    "<NSSystemAppearance: 0x6000008ad680>"
)>
default	11:04:02.994508+0100	Amethyst	Registering for test daemon availability notify post.
default	11:04:02.994676+0100	Amethyst	notify_get_state check indicated test daemon not ready.
default	11:04:02.998620+0100	Amethyst	SignalReady: pid=43913 asn=0x0-0x2c82c8
default	11:04:02.999499+0100	Amethyst	SIGNAL: pid=43913 asn=0x0x-0x2917064
default	11:04:03.481717+0100	Amethyst	*** -[NSBundle initWithURL:]: nil URL argument
default	11:04:03.481935+0100	Amethyst	        OSErr AERemoveEventHandler(AEEventClass, AEEventID, AEEventHandlerUPP, Boolean)(spec,phac handler=0x7fff3381efb6 isSys=YES) err=0/noErr
```

**Comment by ajorpheus** [comment]
Compare the above with the log generated when v0.15.4 starts:

```
default	11:04:02.799417+0100	Amethyst	FRONTLOGGING: version 1
default	11:04:02.799443+0100	Amethyst	Registering, pid=43913
default	11:04:02.800772+0100	Amethyst	CHECKIN: pid=43913
default	11:04:02.814129+0100	Amethyst	CHECKEDIN: pid=43913 asn=0x0-0x2c82c8 foreground=0
default	11:04:02.887854+0100	Amethyst	Registered, pid=43913 ASN=0x0,0x2c82c8
default	11:04:02.887974+0100	Amethyst	Registered, pid=43913 cgConnectionID=c5827
default	11:04:02.889174+0100	Amethyst	BringForward: pid=43913 asn=0x0-0x2c82c8 bringForward=0 foreground=0 uiElement=1 launchedByLS=1 modifiersCount=1 allDisabled=0
default	11:04:02.889214+0100	Amethyst	BringFrontModifier: pid=43913 asn=0x0-0x2c82c8 Modifier 0 hideAfter=0 hideOthers=0 dontMakeFrontmost=0 mouseDown=0/0 seed=0/0
default	11:04:02.891904+0100	Amethyst	Current system appearance, (HLTB: 1), (SLS: 0)
default	11:04:02.897026+0100	Amethyst	Post-registration system appearance: (HLTB: 1)
default	11:04:02.922457+0100	Amethyst	NSApp cache appearance:
-NSRequiresAquaSystemAppearance: 0
-appearance: (null)
-effectiveAppearance: <NSCompositeAppearance: 0x6000008a1680
 (
    "<NSAquaAppearance: 0x6000008ad700>",
    "<NSAquaAppearance: 0x6000008ad600>",
    "<NSSystemAppearance: 0x6000008ad680>"
)>
default	11:04:02.994508+0100	Amethyst	Registering for test daemon availability notify post.
default	11:04:02.994676+0100	Amethyst	notify_get_state check indicated test daemon not ready.
default	11:04:02.998620+0100	Amethyst	SignalReady: pid=43913 asn=0x0-0x2c82c8
default	11:04:02.999499+0100	Amethyst	SIGNAL: pid=43913 asn=0x0x-0x2917064
default	11:04:03.481717+0100	Amethyst	*** -[NSBundle initWithURL:]: nil URL argument
default	11:04:03.481935+0100	Amethyst	        OSErr AERemoveEventHandler(AEEventClass, AEEventID, AEEventHandlerUPP, Boolean)(spec,phac handler=0x7fff3381efb6 isSys=YES) err=0/noErr

```

Note that the following exception in the log from 0.15.6 startup is missing in 0.15.4:
```
default	11:04:03.481717+0100	Amethyst	*** -[NSBundle initWithURL:]: nil URL argument
default	11:04:03.481935+0100	Amethyst	        OSErr AERemoveEventHandler(AEEventClass, AEEventID, AEEventHandlerUPP, Boolean)(spec,phac handler=0x7fff3381efb6 isSys=YES) err=0/noErr
```

So am guessing for some reason, 0.15.6 fails to register an event handler? Which would explain why it does not do anything.

**Comment by ajorpheus** [comment]
I built Amethyst locally from the development branch and installing the app built by Xcode seems to be working fine.

I had a quick look at the commits between 0.15.6 and development (https://github.com/ianyh/Amethyst/compare/v0.15.6...development) but nothing stands out to me.

So at this point, I am not sure where the problem is.

**Comment by ianyh** [comment]
Can you try including the `--include-apps` option to the debug info call? I'm wondering if there's a specific app causing the issue.

**Comment by ajorpheus** [comment]
```
/Applications/Amethyst.app/Contents/MacOS/Amethyst --debug-info --logs --include-apps                                                                                                                                                                 <sandbox-ss>[23-May 15:44:16]
Version: 0.15.6 (99)

OS version: Version 10.15.7 (Build 19H2)

Screens:
	(0.0, 0.0, 2560.0, 1440.0) [(0.0, 0.0, 2560.0, 1440.0)]

Configuration:
debug-layout-info: 1
screen-padding-top: 0
layouts: (
    bsp,
    fullscreen,
    bsp
)
enables-layout-hud-on-space-change: 0
window-minimum-width: 0
enables-layout-hud: 1
floating: (
)
new-windows-to-main: 0
follow-space-thrown-windows: 0
mouse-swaps-windows: 1
mouse-follows-focus: 0
mouse-resizes-windows: 1
float-small-windows: 1
screen-padding-left: 0
window-minimum-height: 0
focus-follows-mouse: 0
window-margins: 0
mod2: (
    option,
    shift,
    control
)
floating-is-blacklist: 1
smart-window-margins: 1
screen-padding-bottom: 0
screen-padding-right: 0
window-margin-size: 8
use-canary-build: 1
window-resize-step: 5
restore-layouts-on-launch: 1
ignore-menu-bar: 0
mod1: (
    option,
    shift
)

Manageable applications:
	talagent (com.apple.talagent)
	Finder (com.apple.finder)
	AppSSOAgent (com.apple.AppSSOAgent)
	Viscosity (com.viscosityvpn.Viscosity)
	Cardhop (com.flexibits.cardhop.mac)
	ToothFairy (com.robinlu.mac.Tooth-Fairy)
	1Password 7 (com.agilebits.onepassword7)
	storeuid (com.apple.storeuid)
	iTerm2 (com.googlecode.iterm2)
	Proxifier (com.initex.proxifier.macosx)
	Google Chrome (com.google.Chrome)
	coreautha (com.apple.LocalAuthentication.UIAgent)
	PhraseExpress (com.bartelsmedia.PhraseExpressOSX)
	IntelliJ IDEA (com.jetbrains.intellij)
	Slack (com.tinyspeck.slackmacgap)
	Beyond Compare (com.ScooterSoftware.BeyondCompare)
```

**Comment by ajorpheus** [comment]
Thank you very much for taking a look @ianyh 
(Also, thank you for this great app)

```
Manageable applications:
	talagent (com.apple.talagent)
	Finder (com.apple.finder)
	AppSSOAgent (com.apple.AppSSOAgent)
	Cardhop (com.flexibits.cardhop.mac)
	ToothFairy (com.robinlu.mac.Tooth-Fairy)
	1Password 7 (com.agilebits.onepassword7)
	storeuid (com.apple.storeuid)
	iTerm2 (com.googlecode.iterm2)
	Google Chrome (com.google.Chrome)
	coreautha (com.apple.LocalAuthentication.UIAgent)
	PhraseExpress (com.bartelsmedia.PhraseExpressOSX)
	Slack (com.tinyspeck.slackmacgap)
	WhatsApp (desktop.WhatsApp)
```

I would've thought that a specific app would cause Amethyst to not do a great job of arranging windows (or be slow at it). Is it possible that it stops working completely?  I have noticed that I need to relaunch Amethyst frequently to get it to do something with the set of apps that I have open usually.

Also, do you think it is possible that for the same set of 'Manageable applications' v0.15.4 works fine?

**Comment by ajorpheus** [comment]
Also, I did notice that when it works 0.15.6 is appreciably faster that 0.15.4

**Comment by ajorpheus** [comment]
@ianyh , interestingly I can't seem to get 0.15.6 output the debug-info anymore!

```
> /Applications/Amethyst.app/Contents/MacOS/Amethyst --debug-info --include-apps   

2022-05-24 09:28:13.687 Amethyst[37826:4333806] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[NSBundle initWithURL:]: nil URL argument'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff310f1b57 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff69f815bf objc_exception_throw + 48
	2   Foundation                          0x00007fff337566cb -[NSVariableExpression isEqual:] + 0
	3   Foundation                          0x00007fff337565c7 +[NSBundle bundleWithURL:] + 33
	4   Silica                              0x00000001009909c4 -[NSRunningApplication(Silica) isAgent] + 65
	5   Amethyst                            0x00000001003cd929 Amethyst + 153897
	6   Amethyst                            0x00000001003cdfaf Amethyst + 155567
	7   Amethyst                            0x00000001003aa5f0 Amethyst + 9712
	8   libdyld.dylib                       0x00007fff6b129cc9 start + 1
	9   ???                                 0x0000000000000003 0x0 + 3
)
libc++abi.dylib: terminating with uncaught exception of type NSException
[1]    37826 abort      /Applications/Amethyst.app/Contents/MacOS/Amethyst --debug-info --include-app
```

**Comment by Pharos3** [comment]
Same for me 0.15.6 stopped working for the past few days. (macos 12.4)

**Comment by ianyh** [comment]
I think this is the actual issue. https://github.com/ianyh/Silica/commit/5084b7aa0bdd16dfe4a230b1e5fe483462ea87d0

Not checking for a bundle URL before accessing the bundle.

**Comment by ajorpheus** [comment]
Built locally with the following commits from `development` at the HEAD:

```
7148c7b (HEAD -> refs/heads/development, refs/remotes/origin/development, refs/remotes/origin/HEAD) Fix ghost windows which have nil Space value under BSP Layout  (#1249)
32c9702 Upgrade Silica (#1250)
a197e0c Custom Layouts (#1134)
2b3432b Fix swapping with two windows visible for two-pane layout (#1241)
```

Seems to work, but can't be sure yet since past development builds also worked for a bit but stopped working after few hours. Will report back in a few hours/days when/if the development build stops working

**Comment by ajorpheus** [comment]
> Built locally with the following commits from `development` at the HEAD:
> 
> ```
> 7148c7b (HEAD -> refs/heads/development, refs/remotes/origin/development, refs/remotes/origin/HEAD) Fix ghost windows which have nil Space value under BSP Layout  (#1249)
> 32c9702 Upgrade Silica (#1250)
> a197e0c Custom Layouts (#1134)
> 2b3432b Fix swapping with two windows visible for two-pane layout (#1241)
> ```
> 
> Seems to work, but can't be sure yet since past development builds also worked for a bit but stopped working after few hours. Will report back in a few hours/days when/if the development build stops working

Still working fine!

**Comment by ajorpheus** [comment]
> > Built locally with the following commits from `development` at the HEAD:
> > ```
> > 7148c7b (HEAD -> refs/heads/development, refs/remotes/origin/development, refs/remotes/origin/HEAD) Fix ghost windows which have nil Space value under BSP Layout  (#1249)
> > 32c9702 Upgrade Silica (#1250)
> > a197e0c Custom Layouts (#1134)
> > 2b3432b Fix swapping with two windows visible for two-pane layout (#1241)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Seems to work, but can't be sure yet since past development builds also worked for a bit but stopped working after few hours. Will report back in a few hours/days when/if the development build stops working
> 
> Still working fine!

It is still working reliably after 3 days. No frequent intermittent issues like before.

@ianyh  I am tempted to suggest that you were spot on with `32c9702 Upgrade Silica (#1250)` as the reason for this.

I am happy to close this issue.

**Comment by ajorpheus** [comment]
@Pharos3 , @mcranieri   are you able to build the development branch locally and give it a go please? If it doesn't work for you, this issue can be re-opened.

**Comment by ajorpheus** [comment]
Last comment!

The locally built version of Amethyst is still working fine after 6 days of usage, so I am convinced that the issue was most probably fixed by `32c9702 Upgrade Silica (#1250)`
